Core/HLE/sceKernelInterrupt.cpp

```diff
@@ -742,10 +742,169 @@ static int sysclib_memcmp(u32 dst, u32 src, u32 size) {
+			// in case we have a bad fmt string, try not to crash the whole emulator
+			fmt_len++;
+			if(fmt_len == 1024){
+				ERROR_LOG(SCEKERNEL, "sprintf, fmt is longer than 1024");
+				return 0;
+			}
+
+			if(state == 0){
+				if(*c == '%'){
+					cur_fmt = "%";
+					state = 1;
+					bytes_to_read = 0;
+				}else{
+					result.append(1, *c);
+				}
+			}else{
+				cur_fmt.append(1, *c);
+
+				// going by https://cplusplus.com/reference/cstdio/printf/#compatibility
+				// no idea what the kernel module really supports as of writing this
+
+				if(*c == '%'){
+					result.append(cur_fmt);
+					state = 0;
+				}else if(*c == 'd' ||
+					*c == 'i' ||
+					*c == 'u' ||
+					*c == 'o' ||
+					*c == 'x' ||
+					*c == 'X' ||
+					*c == 'f' ||
+					*c == 'e' ||
+					*c == 'E' ||
+					*c == 'g' ||
+					*c == 'G' ||
+					*c == 'c' ||
+					*c == 'p' ||
+					*c == 'n' ||
+					*c == 's'
+				){
+					// we have a format ending
+					if(*c == 's'){
+						// consume 4 bytes from arguments
+						u32 val = 0;
+						if(arg_idx <= 1){
+							val = currentMIPS->r[MIPS_REG_A2 + arg_idx];
+						}else if(arg_idx <= 5){
+							val = currentMIPS->r[MIPS_REG_T0 + arg_idx - 2];
+						}else{
+							// XXX assuming right to left arg pushing
+							int stack_idx = arg_idx - 6;
+							u32 stack_cur = psp_stack_pointer + stack_idx * 4;
+
+							if(!Memory::IsValidAddress(stack_cur)){
+								ERROR_LOG(SCEKERNEL, "sprintf, bad stack pointer %08x", stack_cur);
+								return 0;
+							}
+							val = Memory::Read_U32(stack_cur);
+						}
+						arg_idx++;
+
+						if(!Memory::IsValidAddress(val)){
+							ERROR_LOG(SCEKERNEL, "sprintf, bad string reference %08x", val);
+							return 0;
+						}
+
+						const char *str = Memory::GetCharPointerUnchecked(val);
+						// limit the string length and hope that we don't crash on a bad string reference
+						char buf[1024] = {0};
+						strncpy(buf, str, sizeof(buf));
+						buf[sizeof(buf) - 1] = '\0';
+						result.append(buf);
+					}else{
+						u64 val = 0;
+						if(bytes_to_read == 0){
+							bytes_to_read = 4;
+						}
+						int read_cnt = 0;
+						while(bytes_to_read != 0){
+							u64 val_from_arg = 0;
+							if(arg_idx <= 1){
+								val_from_arg = currentMIPS->r[MIPS_REG_A2 + arg_idx];
+							}else if(arg_idx <= 5){
+								val_from_arg = currentMIPS->r[MIPS_REG_T0 + arg_idx - 2];
+							}else{
+								// XXX assuming right to left arg pushing
+								int stack_idx = arg_idx - 6;
+								u32 stack_cur = psp_stack_pointer + stack_idx * 4;
+
+								if(!Memory::IsValidAddress(stack_cur)){
+									ERROR_LOG(SCEKERNEL, "sprintf, bad stack pointer %08x", stack_cur);
+									return 0;
+								}
+								val_from_arg = Memory::Read_U32(stack_cur);
+							}
+							arg_idx++;
+
+							val = val | (val_from_arg << (read_cnt * 32));
+
+							bytes_to_read = bytes_to_read - 4;
+							read_cnt++;
+						}
+						/*
+						// windows compiler don't like this
```

Comment:
For good reason, dynamic stack allocation is not a good idea.

Reply:
Hm, you're right, I supposed I should not be too trusty with snprintf's return value being always